Add coverage to the project reports.

kinow opened the pull request:
Travis-CI configuration was updated to use build stages. The build matrix was moved to the test stage, and extra configuration for coverage.py and codacy were added.

The tool used to collect coverage from the tests in [Coverage.py](https://coverage.readthedocs.io/en/coverage-4.5.1a/), which is supported by multiple build and reporting systems.

Due to the nature of the tests in cylc, and to how they run in parallel in different containers in Travis-CI, we had to use a [`sitecustomize.py`](https://docs.python.org/3/library/site.html) script.

As Codacy was already used for code analysis, this pull request includes settings to export the coverage report to Codacy. I assume the Codacy token is already set up. So maybe just merging the pull request would already make the coverage report available in the [cylc dashboard](https://app.codacy.com/project/Cylc/cylc/dashboard).

It is also possible to get the coverage of pull requests.

Cheers
Bruno

kinow commented:
The branch from where this pull request is coming from has coverage already in Codacy. The Travis-CI build can be seen here: [https://travis-ci.org/kinow/cylc](https://travis-ci.org/kinow/cylc)

And the Codacy report is also available here: [https://app.codacy.com/project/kinow/cylc/dashboard?branchId=8624609]

And here are some screenshots: 

![screenshot_2018-08-31_00-37-31](https://user-images.githubusercontent.com/304786/44894782-a5c06c00-ad45-11e8-9d76-ad7920c72dbc.png)

![screenshot_2018-08-31_17-26-17](https://user-images.githubusercontent.com/304786/44894783-a5c06c00-ad45-11e8-8815-104707f698a9.png)

![screenshot_2018-08-31_17-26-43](https://user-images.githubusercontent.com/304786/44894784-a6590280-ad45-11e8-9db9-8623d03168ea.png)

![screenshot_2018-08-31_17-34-54](https://user-images.githubusercontent.com/304786/44894785-a6590280-ad45-11e8-953c-a2da0331ab71.png)

![screenshot_2018-08-31_17-35-13](https://user-images.githubusercontent.com/304786/44894786-a6590280-ad45-11e8-8d6f-8617e3ab3745.png)

kinow commented:
Codacy is unhappy with the use of `subprocess` in the `cover.py` script. That script is used to invoke the tests. Previously, the test command was in `.travis-ci.yml`. But now we have to start it with `coverage run`, which only supports Python scripts.

So that script is just a Python wrapper, so that `coverage run` works normally.

kinow commented:
Added a commit with the coverage badge, and mentioned the issue #2765 too.

I believe someone with karma will have to set up the Codacy token in order for the coverage to work in Codacy.

The [first build in Travis for this pull request](https://travis-ci.org/cylc/cylc/builds/422881118) tried to upload the coverage to the cylc/cylc Codacy dashboard, but it failed with

```
08/31 06:08:11 [1;31mERROR[0;39m [36mc.c.rules.ConfigurationRules:22[0;39m - Invalid configuration: Project token not provided and not available in environment variable "CODACY_PROJECT_TOKEN" 
```

At least Travis-CI build is still working :-)

hjoliver commented:
@cylc/core - note you can drill down (from the page shown in the last of Bruno's screenshots above) to see exactly what lines of code are touched by the test battery.  Coverage does not look as good as I had expected/hoped.  Hopefully our heavy use of full regression tests means that code associated with "normal" Cylc usage is well covered, but maybe we are missing a lot of edge cases (which may be difficult to target with regression tests in a vanilla test environment). 

So perhaps this warrants a change to our testing strategy?  How about, for all new features and fixes:
- write new (or modify existing) regression tests to support major functionality
- guided by coverage data, write unit tests (where possible) to nail any code that is not touch by the regression tests.
- (and we can incrementally retro-fit more unit tests to existing files touched by new PRs)

Perhaps we should consider full coverage by unit tests (where possible) but my feeling is that would be a lot of redundant effort given that we have to have regression tests as well, to ensure that the functionality actually relied on by users works.

@kinow might have opinions on this?

hjoliver commented:
Pinged @oliver-sanders as first reviewer, since this remodels his recent "chunking" of the test battery (if he has time pre-Brussels).

kinow commented:
@hjoliver ,

>Coverage does not look as good as I had expected/hoped. Hopefully our heavy use of full regression tests means that code associated with "normal" Cylc usage is well covered, but maybe we are missing a lot of edge cases (which may be difficult to target with regression tests in a vanilla test environment).

That's my thinking as well. I want to keep running the coverage locally, experimenting with the settings, and confirming that it is not missing any tests when reporting the coverage.

There are many cases - especially exception handling - where I think it could be easier to be covered by unit tests (and some times simple mocks). Not necessarily to increase the coverage, but more to make sure the logic behind the exception handling is working as expected. Same for validation.

I haven't checked, but if I remember well, we have some files in Cylc's folder with both the main code, and the unit test. Might be worth checking if these are not interfering with the reports (and in that case, it may be a good idea to move the test to its own folder).

Happy to take more reviews/suggestions/comments :-)

Thanks!
Bruno

hjoliver commented:
@kinow - for the unit tests that we do have (admittedly not so many, as discussed!) I believe we run them as special tests within `cylc test-battery`, so should not be missed in the coverage reports.

kinow commented:
Running a small experiment, where I've added a commit with the `unitest.TestCase` instances moved to a different package. Tried to match settings in Travis-CI and Coverage.py commands, in order to get a better coverage number... now wait and see if Travis-CI will run successfully :-)

kinow commented:
If I am reading the Codacy user interface correctly, coverage increased 4%

![image](https://user-images.githubusercontent.com/304786/45214014-61a50c80-b2ed-11e8-8b4b-96469c6f5aee.png)

Now omitting the `isodatetime` folder too, as I had a look, and it looks like this folder is a copy of the `metomi/isodatetime` module.

kinow commented:
Done! The current Travis-CI job for Cylc doesn't have the Codacy integration enabled, but I have enabled Travis-CI and Codacy for my fork. So the latest coverage report (as well as the coverage trend) can be seen here:

- https://travis-ci.org/kinow/cylc
- https://app.codacy.com/project/kinow/cylc/dashboard

After omitting `isodatetime`, coverage went up to **42%**. I think `xdot.py` and `cylc_xdot.py` can be possibly omitted from the reports too, but quite happy with the current setup.

![image](https://user-images.githubusercontent.com/304786/45214402-7f26a600-b2ee-11e8-9269-9fb13a5536f5.png)

Unit tests would be under the `lib/cylc/tests` folder now, and executed with PyTest. Integration/functional tests would be under the `tests/` folder, and keep being executed with Shell/Python/prove. The report that we have is an aggregated report.

There are a few files that have 0% coverage, but contain Python 2 statements that would have to be ported to Python 3. So keen to start writing some unit/functional tests to cover the current behaviour too, and try to reduce the likelihood of regressions (that may help the work on the new Web GUI as well I think).

Bruno

hjoliver commented:
Nice. Yeah we should definitely look at those 0% files and write some tests.  I'll try to take a detailed look at the report soon.

hjoliver commented:
@kinow - we took a look at this yesterday and we're generally impressed at how useful the coverage reports will be.  We might as well merge it and start using it, I think.   Is this PR good to go as far as you're concerned, once https://github.com/cylc/cylc/pull/2766#issuecomment-417563008 is done?

kinow commented:
Hi @hjoliver , I rebased it this week, and had another look at the changes, and so far everything looks OK to me. The tests appear to be working OK in Travis-CI, and the report also appears to be generated correctly.

So +1, if someone can look at the settings for Codacy, then I think this PR is good to go :-) Hopefully this report - as well as the other reports/reviews - will support us during the upcoming changes for web gui, python3 :tada: 

Thank you

oliver-sanders commented:
FYI more unittests have snuck in (`scheduler_cli`)

hjoliver commented:
@kinow - should I be able to see a coverage report now? (can't find it).

kinow commented:
@hjoliver I was scratching my head and trying to understand why the coverage was not uploaded, though the tests were run successfully.

Well, I copied most of the new settings from isodatetime, where tests run with Python 2.6 and 2.7, but coverage only runs when a `RUN_COVERAGE` shell flag is enabled.

I forgot to remove the check for that flag 😅 done, so hopefully next run will give us the new numbers in Codecov. Thanks!

kinow commented:
Thanks @oliver-sanders ! grep'ed to search other tests, but we had only the `scheduler_cli` new tests. Moved to `/tests/`, and updated branch (will rebase/squash/etc later).

kinow commented:
@hjoliver success! (apparently) 🎉🚀 

https://codecov.io/gh/cylc/cylc

We are now at near 59% coverage, which is impressive given the size and complexity of the project, but there are still parts that will need to be changed for Python 3 but are not yet covered 😟

kinow commented:
Rebased onto tip of master, then moved commits around, squashed them, and got down to just 2 commits.

Pull request is now ready for review again.

kinow commented:
Note: there is no Codecov under the checks in this pull request, even though the coverage was successfully reported to Codecov, and can be seen online already.

The reason for that is that in order to have the Codecov Checks with links to the report in the pull request/branch, we would have to have a webhook.

https://github.com/apps/codecov installing this github app, and selecting the cylc repository, automatically creates the webhooks.

hjoliver commented:
@kinow - I think I've authorized codecov on cylc/cylc following  your link above.  Can you push some trivial change to your branch to trigger regeneration of the checks?

kinow commented:
Sure thing @hjoliver . I had actually accidentally added a comment to `.travis.yml` and avoided removing it to prevent yet another build in travis. So submitted an extra commit to remove that comment... let's see if Codecov will appear as a check here now.

hjoliver commented:
Woo hoo, there it is :boom:  Nice.

kinow commented:
Hooray!!! 🎉 💥 (stealing the emoji from @hjoliver)

I noticed the threshold is still set for the same amount as in isodatetime.

isodatetime initially had something around 85% or more of coverage I think. So I set a threshold of 75% to let the team later decide what that number should look like.

Cylc is on 59.4%. Should we set the passing bar at 55% for now? And slowly increase? Without doing that, Codecov will fail every pull request... which I'm fine with, but not sure if that'd be distracting to others?

kinow commented:
And I believe the Codacy issues here can be all ignored, as they happen in the new Python script used exclusively for Travis-CI. I can actually exclude it in Codacy if others are Ok with that, so that Codacy doesn't keep popping up more comments in pull requests.

hjoliver commented:
I'm happy to exclude that from codacy.

kinow commented:
Just realized I didn't include the headers in the test files. So another update for this pull request in some hours :+1:

hjoliver commented:
> @hjoliver I think you meant LGTM (is it still trendy if you get it wrong? laughing)

God, how humiliating! :sob:

oliver-sanders commented:
@kinow Branch in conflicts :cry:, happy to merge once resolved.

kinow commented:
@oliver-sanders just rebased and fixed the conflict. It was from extra tests which I added for the `GraphParser` in another pull request.

I also added one more last minute change to the pull request (hope that's OK). Actually two. One was a call to `coverage combine` to merge all `.coverage...` files. It was happening anyway, but only because Codecov's shell script apparently does that by default. This way I think it's safer in case we move to another coverage tool or if Codecov's code change.

The second change is also simple... I moved the unit tests to before the functional tests. Interestingly after I did that, it complained about the `mock` module missing (thought Travis included it by default for Python 2.7), so I also added that module (already documented in `cug.tex`).

The reason for this last minute change, is mainly because unit tests are cheaper (both to troubleshoot and to run) than functional tests. So in case we have a change that breaks unit tests, the build should fail sooner, without bothering to start the functional tests :star: 

@hjoliver I think I mentioned in a previous comment we had around 58%... after I rebased and incorporated the latest changes, it went down to around 55%. Still within the target of 58% that we set before with a threshold of 5%.

Ready for another review :+1: 

Thanks
Bruno

hjoliver commented:
> after I rebased and incorporated the latest changes, it went down to around 55%. 

OK. Once this is merged, we can start making sure coverage only ever goes up.

oliver-sanders commented on .travis/sitecustomize.py:
```diff
@@ -0,0 +1,23 @@
+# Cylc and Python aware of it, we export PYTHONPATH when running the tests
+
+import coverage
+coverage.process_startup()
```
What calls this `sitecustomize.py` script?

kinow replied:
We are setting the `PYTHONPATH` variable in the `.travis.yml` script. When Python finds that script in the `PYTHONPATH`, it loads it (similarly to those `.pth` files). Here's the recent docs, but that applies to Python 2.x too : https://docs.python.org/3/library/site.html (search for sitecustomize).

Handy trick, that we had to use because the coverage module needs to be aware that we want it to run even though what started it was a shell script in the functional tests with `test-battery`.

oliver-sanders replied:
Cool, diddn't know about that one!

oliver-sanders commented on lib/cylc/tests/test_param_expand.py:
```diff
@@ -0,0 +1,209 @@
+        """Test graph expansion with two parameters each side of an arrow."""
+        self.assertEqual(
+            self.graph_expander.expand("bar<i,j>=>baz<i,j>"),
+            {"bar_i0_j1=>baz_i0_j1",
```
Sadly we can't get away with Python 3 syntax just yet! See https://github.com/kinow/cylc/pull/1.

kinow replied:
Oh, sorry, I tested only on 2.7! Will have to remember the paragraph from our user guide 

>Python 2 >= 2.6 is required. Python 2 >= 2.7.9 is recommended for the best security. Python 2 should already be installed in your Linux system. https://python.org/. 

Merging it and updating this pull request.

kinow commented on .travis.yml:
```diff
@@ -70,15 +70,26 @@ install:
-    # Run tests with virtual frame buffer for X support.
-    - xvfb-run -a cylc test-battery --chunk $CHUNK --state=save -j 5 || (echo -e "\n\nRerunning Failed Tests...\n\n"; cylc test-battery --state=failed -j 5)
+    - export COVERAGE_PROCESS_START="${TRAVIS_BUILD_DIR}/.coveragerc"
+    - PYTHONPATH=$(pwd -P)/lib/ py.test --cov=lib/cylc lib/cylc/tests/ --cov-append
```
Oh, I think I see what you mean. There's no point in running the unit tests 4 times.

As the `COVERAGE_PROCESS_START` is necessary for the other tests, I'm moving it under `install` section (or can move to `before_install` if that makes more sense).

kinow commented:
Rebased onto master just to check current coverage on master :+1:

oliver-sanders commented:
Heads up #2809 will move the `HostAppointer` class (along with its unittests) into `cylc.host_appointer` (sorry)!

kinow commented:
Thanks @oliver-sanders ! Adding another post-it on my monitor so I don't forget it :nerd_face: Thanks!

matthewrmshin commented:
This can probably go into the next release? (Low risk changes?)

matthewrmshin commented:
Conflict.

kinow commented:
_Conflict resolution level-up!_ :star: :video_game:

hjoliver commented:
@matthewrmshin @oliver-sanders - I'm about to go off-line for my trip; but can you consider promoting this to next-release? (I'm not aware of any reason to hold it back at this point, and it'll be good to have coverage on subsequent PRs...)

matthewrmshin commented:
Agreed that this should go into the next release.

matthewrmshin commented:
#2809 and this will conflict (and may require some post-merge fixes). @oliver-sanders @kinow (and @hjoliver) Please agree on which one to merge first.

kinow commented:
@matthewrmshin, I am following the work on that auto stop-restart pull request. Really interesting! And also looks much harder than this one.

@oliver-sanders happy to wait till that one is done to rebase this one, as I believe it will be much easier for me to simply update the travis/coverage files, and have another look around for any `unittest` that may have slipped into the source folders :+1:

oliver-sanders commented:
@kinow If you are OK with that, great. I think this is all ready to go in.

kinow commented:
>Looking forward to having coverage reports!

+1 @oliver-sanders  🎉 !

In the meantime I will polish this pull request a bit more.

Just add to the ignore list the files related to profiling, GUI, and the ones that will be deprecated/removed when we move to Python3 (e.g. parsec's OrderedDictCompat).

Coverage now up to 60.81%. Including the parsec unit tests in the other pull request, I guess we should be around 65% now. And this week I am planning to finish another batch of unit tests. Aiming at near 75% pre-Melbourne workshop (which I already have tickets/accommodation/visa/all sorted 🎉, thanks to @hjoliver !)

kinow commented:
Correction, enabled all doctest tests in pytest (removing any functional test that was triggering doctest, which should speed up a tiny bit the build on Travis).

With that, the coverage went up to **63.03%**. With the new parsec tests in the other pull request, I believe we should be getting close to 70% by the end of this week with the extra tests! 🎉 🎈

matthewrmshin commented on lib/cylc/tests/__init__.py:
```diff
@@ -0,0 +1,76 @@
+        try:
+            importlib.import_module(module_name)
+        except ImportError as exc:
+            print 'doctest: skipping "%s" - %s' % (module_name, exc)
```
Is this just for user diagnostics?
* If so, should probably go to STDERR? And should we use the print function instead of the the print statement?
* Is there a better way to alert users that we are skipping tests? (Like we have in TAP format?)

oliver-sanders replied:
* This message is of the form "skipped python module not available" which is roughly comparable to the "skipped: command not available" messages we get from the TAP test harness. In TAP these messages are printed to stdout no stderr. Can convert to print functions if desired.
* In the unittest framework tests can be skipped but they must first exist in order to be skipped. The test suite is created here: https://github.com/cylc/cylc/pull/2766/files/6b3d4a82420a82e0424f8b79403708d9ce39374d#diff-d15de699c9b87f6274e3824b0b02abb2R56

kinow replied:
@oliver-sanders I think that code can now be deprecated, as the `doctest`s are now running with PyTest.

Checked running `py.test` with and without this code, and the result was the same.

Is it something that we need to keep for Python 2.6 compatibility? Just committed the change removing it to trigger Travis and check the coverage/test reports, but happy to revert if necessary 👍

oliver-sanders replied:
Always remove if not necessary!

> "I hate code, and I want as little of it as possible in our product." - Jack Diederich

So long as py.test picks up any and all doctests that's great!

kinow commented:
Added `mkdir_p.py` to the ignore list, see #2858

oliver-sanders commented on .travis.yml:
```diff
@@ -31,6 +31,9 @@ matrix:
       env: CHUNK='1/4'
+      before_script:
+        # we must run unit tests only once
+        - PYTHONPATH=$(pwd -P)/lib/ py.test
```
> Are you running py.test or pytest

`pytest` (we do not have access to the older `py.test`).

> in the master branch

In `kinow/report-test-coverage` (7.7.1-389-g578eeee04).

hjoliver replied:
(I believe this discussion has been resolved/explained by @matthewrmshin)

kinow commented:
I think it will be easier to remove it from next-release and move to later/some day/etc.

Maybe after or during Py3. As it would have less envs to validate against, and we'd all have a similar env (including travis).

I lost momentum and now going through CUG yet again, and reviewing things for Melbourne.

Cheers

p.s. it is possible to branch from this ranch, amd create temp branches to see the coverage of pull requests, etc, in the meantime. In case anyone needs it now :+1:

sadielbartholomew commented:
> I think it will be easier to remove it from next-release and move to later/some day/etc.

@kinow this seems a shame, since it will be really useful & you have clearly put a lot of effort into it. I would be happy to take over this from you, especially being on the Met Office side where the testing complications lie, should:
* you be happy for me to do so;
* you & the rest of @cylc/core think that it would be possible (clearly with a significant amount of exploration &/or trial & error with setup) & worthwhile in context of the impending Dev Workshop to get the tests to run &/or skip them as necessary for our site?

matthewrmshin commented:
In my work site environment, I am unable to get `pytest` to understand the `-cov` option in the `pytest.ini` file. I can run the test if I temporarily move `pytest.ini` to one side. Is there a better solution? Again, happy to live with this for a short time until we get the packaging solution sorted.

kinow commented:
Oh, I think that's another mistake I made in `pytest.ini`. I instructed it to always include the coverage parameters.

`--cov` and related coverage parameters are not supported in PyTest. Instead, it is provided only via plugins (which use `setup.py` + entrypoints :tada: , which we could also use to replace jinja filters, parsec config parsers, etc).

Moved the coverage specific parameters to the `.travis.yml` file. Now running `pytest` should work. Also replaced `py.test` by `pytest` in Travis, to use the newer tool :+1: 

Rebased branch, Travis will be busy for a while building it now. Thanks Matt!

hjoliver commented:
@kinow - I think I might have just conflicted this branch by merging the auto stop-restart one :disappointed:

kinow commented:
Not a problem @hjoliver , conflicts resolved. Travis is building now. Thanks for the heads up! :+1:

kinow commented:
@hjoliver pointed in #2877 that some unit tests were still executed with the `test-battery` command. Some were not even being executed anymore, and the `unitttest` test class had already been moved. So I've removed the tests that pytest was already taking care of.

Also include an extra commit to disable a `doctest` from parsec. That is already executed by pytest, and included in the coverage reports :+1:

kinow commented:
Replaced [deprecated](https://docs.python.org/2/library/unittest.html#deprecated-aliases) `failUnless` by `assertTrue` in the last commit, to make a few warnings go away when running `pytest`.

matthewrmshin commented:
This is now looking good.

In my environment, I need to exclude these for lacking GTK and EmPy:

```diff
diff --git a/pytest.ini b/pytest.ini
index 3a8ae0802..c9e260686 100644
--- a/pytest.ini
+++ b/pytest.ini
@@ -17,11 +17,14 @@
 [pytest]
 addopts = --verbose
     --doctest-modules
+    --ignore=lib/parsec/empysupport.py
     --ignore=lib/parsec/tests/getcfg/bin/one-line.py
     --ignore=lib/parsec/tests/synonyms/bin/synonyms.py
     --ignore=lib/parsec/tests/nullcfg/bin/empty.py
     --ignore=lib/parsec/example
     --ignore=lib/cherrypy
+    --ignore=lib/cylc/cylc_xdot.py
+    --ignore=lib/cylc/gui
     --ignore=lib/isodatetime
     --ignore=lib/markupsafe
 testpaths =
```

I also got some deprecation warnings, because we are using `.failUnless` in
`lib/cylc/tests/test_host_appointer.py`.

kinow commented:
@matthewrmshin are you still have problems with the `test_empysupport.py`? I thought my fix would work on your environment too.

And as for the errors with GTK and xdot... what if we add these lines to `pytest.ini` Matt? These files are not covered by unit tests (as far as I recall). And are also ignored in our coverage report anyway as far as I know.

So that way it would work for everybody with `pytest` :grimacing:

matthewrmshin commented:
Yes, would be useful to ignore the GTK stuffs.

kinow commented:
@oliver-sanders 

>https://pypi.org/project/pytest-parallel/ - later on if our unittests start taking more than a few seconds?

+1

>Better system for automatically skipping unit-tests based on software dependencies. The current system in the cylc test-battery works well, manually editing the .pytest file isn't great.

Once we have `pytest` configured and running, we can start using [markers](https://docs.pytest.org/en/latest/example/markers.html).

>Bring together the pylint and cylc test-battery commands somehow. For example the following change to cylc test-battery will run pytest and pycodestyle before functional tests unless a test or list of tests is explicitly provided (e.g. by cylc test-battery --chunk):

I prefer to leave the test-battery responsible for running the functional tests with shell + Python for Cylc. And have pytest running the tests. `setup.py` and `tox` both interact with `pytest`, none with the `test-battery`. If we adopted `tox` to validate our `setup.py` against multiple Python versions (e.g. 3.3, 3.5, 3.7, etc), we would have to keep changing the test-battery to accommodate these changes.

But we can discuss that later in Melbourne or in cylc/admin, etc. :+1:

matthewrmshin commented:
I am going to merge this and raise a further PR to modify `pytest.ini` so it works at our site.

oliver-sanders commented:
Fine by me
